### PR TITLE
Support OpenSSL on Windows

### PR DESCRIPTION
**Pull request by HertzDevil**
This PR adds OpenSSL 3.0.0 to Windows CI as a static library. Unlike on Linux, we do not have `pkg-config` on Windows, but we do have full control over the exact libraries used by Crystal, so this PR simply ~~outputs the OpenSSL version number somewhere in `Crystal::LIBRARY_PATH` as part of the setup process, then uses the `read_file?` macro method to retrieve it at `LibCrypto` and `LibSSL`~~ retrieves the version number from the `CRYSTAL_OPENSSL_VERSION` environment variable.

Building OpenSSL locally in the same manner as CI requires [Strawberry Perl](https://strawberryperl.com/) and [NASM](https://www.nasm.us/). Although this PR doesn't have it, I have also tried LibreSSL myself and everything works too (that one uses CMake and the Microsoft assembler).

`OPENSSLDIR` is left intact (`C:\Program Files\Common Files\SSL`).

This enables 295 passing and 2 pending examples, such as the remaining digest algorithms, in the standard library test suite.

### CODE REVIEW

File: src/openssl/lib_crypto.cr [hunk 1]
```diff
@@ -1,9 +1,23 @@
...
+      {% for dir in Crystal::LIBRARY_PATH.split(';') %}
+        {% unless ssl_version %}
+          {% config_path = "#{dir.id}\\openssl_VERSION" %}
+          {% if config_version = read_file?(config_path) %}
+            {% ssl_version = config_version.chomp %}
+          {% end %}
+        {% end %}
+      {% end %}
+      {% ssl_version ||= "0.0.0" %}
+    {% else %}
```
Comment (straight-shoota) [inline src/openssl/lib_crypto.cr:15]: I'd suggest to keep this complex macro logic out of this PR for now. As a minimal solution, we can just use an environment variable for now. That's sufficient for CI.

Storing the version information in a file is a distribution feature. How we implement such configuration should be discussed separately.

File: src/http/client/response.cr [hunk 2]
```diff
@@ -1,154 +1,156 @@
...
-  def initialize(@status : HTTP::Status, @body : String? = nil, @headers : Headers = Headers.new, status_message = nil, @version = "HTTP/1.1", @body_io = nil)
-    @status_message = status_message || @status.description
-
-    if Response.mandatory_body?(@status)
-      @body = "" unless @body || @body_io
-    else
-      if (@body || @body_io) && (headers["Content-Length"]? != "0")
-        raise ArgumentError.new("Status #{status.code} should not have a body")
+class HTTP::Client
+  class Response
```
Comment (straight-shoota) [inline src/http/client/response.cr:5]: Why this change? It seems completely unrelated.

Comment (HertzDevil) [reply src/http/client/response.cr:5]: It is necessary because `win32_std_spec.cr` now requires this file from `http/common.cr`, and without this `HTTP::Client` would be first declared as a module which breaks further requires.

Comment (HertzDevil) [reply src/http/client/response.cr:5]: By the way, similar namespaced types under `HTTP::Server` do this already.

File: .github/workflows/win.yml [hunk 3]
```diff
@@ -138,6 +144,41 @@ jobs:
...
+        id: cache-openssl
+        uses: actions/cache@v2
+        with:
+          path: |
+            libs/crypto.lib
+            libs/ssl.lib
+          key: win-openssl-libs-3.0.0
+      - name: Set up NASM
+        if: steps.cache-openssl.outputs.cache-hit != 'true'
+        uses: ilammy/setup-nasm@v1.2.1
```
Comment (oprypin) [inline .github/workflows/win.yml:157]: Please pin the commit for security
```suggestion
        uses: ilammy/setup-nasm@e2335e5fc95548c09cd2deea2768793e0e8f0941 # v1.2.1
```

File: src/openssl/lib_crypto.cr [hunk 4]
```diff
@@ -15,7 +21,13 @@
   end
 {% end %}
 
-@[Link(ldflags: "`command -v pkg-config > /dev/null && pkg-config --libs --silence-errors libcrypto || printf %s '-lcrypto'`")]
+{% if flag?(:win32) %}
+  @[Link("crypto")]
+  @[Link("crypt32")] # CertOpenStore, ...
+  @[Link("user32")]  # GetProcessWindowStation, GetUserObjectInformationW, _MessageBoxW
```
Comment (oprypin) [inline src/openssl/lib_crypto.cr:27]: Linking here seems unfortunate, particularly for `_MessageBoxW` :/
But seems like it couldn't be declared within OpenSSL itself? 
https://github.com/openssl/openssl/blob/3e3ad3c54855dd534437871a0c78858de5e3d246/NOTES-WINDOWS.md?plain=1#L207

**Review by oprypin — APPROVED** [review]

**Review by straight-shoota — APPROVED** [review]

File: src/openssl/lib_crypto.cr (hunk 1 above)
Comment (HertzDevil) [reply src/openssl/lib_crypto.cr]: #11492
